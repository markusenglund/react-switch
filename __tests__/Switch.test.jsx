import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Switch from "../src";

const noop = () => {};

describe("Props", () => {
  it("matches snapshot with default props", () => {
    const { container } = render(<Switch checked={false} onChange={noop} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("matches snapshot with custom props", () => {
    const { container } = render(
      <Switch
        checked
        onChange={noop}
        disabled
        onColor="#abc"
        offHandleColor="#def"
        handleDiameter={9}
        checkedIcon={
          <div
            style={{
              color: "pink",
              paddingTop: 2,
              paddingLeft: 6,
              fontSize: 20
            }}
          >
            ❤
          </div>
        }
        boxShadow="0px 1px 5px rgba(0, 0, 0, 0.6)"
        height={13}
        width={37}
        className="qux"
        id="foo"
        aria-labelledby="baz"
        aria-label="bar"
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("matches snapshot with horizontal switch", () => {
    const wrapper = render(
      <Switch
        checked
        onChange={noop}
        disabled
        onColor="#abc"
        offHandleColor="#def"
        handleDiameter={9}
        checkedIcon={
          <div
            style={{
              color: "pink",
              paddingTop: 2,
              paddingLeft: 6,
              fontSize: 20
            }}
          >
            ❤
          </div>
        }
        boxShadow="0px 1px 5px rgba(0, 0, 0, 0.6)"
        height={37}
        width={13}
        className="qux"
        id="foo"
        aria-labelledby="baz"
        aria-label="bar"
      />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

describe("handle", () => {
  it("sets a boxShadow on focus and loses it on blur", () => {
    const { container } = render(<Switch onChange={noop} checked={false} />);
    fireEvent.focus(container.querySelector("input"));
    expect(
      container
        .querySelector(".react-switch-handle")
        .style.getPropertyValue("box-shadow")
    ).toBe("0 0 2px 3px #3bf");
    fireEvent.blur(container.querySelector("input"));
    expect(
      container
        .querySelector(".react-switch-handle")
        .style.getPropertyValue("box-shadow")
    ).toBe("");
  });

  it("doesn't call onChange on some other key", () => {
    const mockOnChange = jest.fn();
    const { container } = render(
      <Switch onChange={mockOnChange} checked={false} />
    );
    fireEvent.focus(container.querySelector("input"));
    fireEvent.keyDown(container.querySelector("input"), {
      keyCode: 14,
      preventDefault: noop
    });
    expect(mockOnChange).not.toBeCalled();
  });
});

describe("checked prop", () => {
  it("affects transform style when it changes", () => {
    const { container, rerender } = render(
      <Switch onChange={noop} checked={false} />
    );
    rerender(<Switch onChange={noop} checked />);
    expect(
      container
        .querySelector(".react-switch-handle")
        .style.getPropertyValue("transform")
    ).toBe("translateX(29px)");
    rerender(<Switch onChange={noop} checked={false} />);
    expect(
      container
        .querySelector(".react-switch-handle")
        .style.getPropertyValue("transform")
    ).toBe("translateX(1px)");
  });

  it("affects background when it changes", () => {
    const { container, rerender } = render(
      <Switch
        onChange={noop}
        checked={false}
        offColor="#aabbcc"
        onColor="#ddeeff"
      />
    );
    rerender(
      <Switch onChange={noop} checked offColor="#aabbcc" onColor="#ddeeff" />
    );
    expect(
      container
        .querySelector(".react-switch-bg")
        .style.getPropertyValue("background")
    ).toBe("rgb(221, 238, 255)");
    rerender(
      <Switch
        onChange={noop}
        checked={false}
        offColor="#aabbcc"
        onColor="#ddeeff"
      />
    );
    expect(
      container
        .querySelector(".react-switch-bg")
        .style.getPropertyValue("background")
    ).toBe("rgb(170, 187, 204)");
  });
});

// Test dragging behaviour
/*
describe("drag related behaviour", () => {
  it("triggers onChange callback when isDragging is never set to true", () => {
    const mockOnChange = jest.fn();
    const wrapper = shallow(<Switch onChange={mockOnChange} checked={false} />);
    wrapper
      .find("input")
      .simulate("touchstart", { touches: [{ clientX: 100 }] });
    wrapper.find("input").simulate("touchend", { preventDefault: noop });
    expect(mockOnChange.mock.calls[mockOnChange.mock.calls.length - 1][0]).toBe(
      true
    );
  });
  it("triggers onChange when handle is dragged half way or more", () => {
    const mockOnChange = jest.fn();
    const wrapper = shallow(
      <Switch onChange={mockOnChange} checked={false} height={15} width={30} />
    );
    wrapper
      .find("input")
      .simulate("touchstart", { touches: [{ clientX: 100 }] });
    wrapper
      .find("input")
      .simulate("touchmove", { touches: [{ clientX: 107.5 }] });
    wrapper.find("input").simulate("touchend", { preventDefault: noop });
    expect(mockOnChange.mock.calls[mockOnChange.mock.calls.length - 1][0]).toBe(
      true
    );
  });
  it("doesn't trigger onChange when handle is dragged less than half way", () => {
    const mockOnChange = jest.fn();
    const wrapper = shallow(
      <Switch onChange={mockOnChange} checked={false} height={15} width={30} />
    );
    wrapper
      .find("input")
      .simulate("touchstart", { touches: [{ clientX: 100 }] });
    wrapper
      .find("input")
      .simulate("touchmove", { touches: [{ clientX: 107.4 }] });
    wrapper.find("input").simulate("touchend", { preventDefault: noop });
    expect(mockOnChange).not.toBeCalled();
  });
});
*/
