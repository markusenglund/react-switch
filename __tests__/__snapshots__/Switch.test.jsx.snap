// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Props matches snapshot with custom props 1`] = `
<div
  class="qux"
  style="position:relative;display:inline-block;text-align:left;opacity:0.5;direction:ltr;border-radius:6.5px;-webkit-transition:opacity 0.25s;-moz-transition:opacity 0.25s;transition:opacity 0.25s;touch-action:none;-webkit-tap-highlight-color:rgba(0, 0, 0, 0);-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none"
>
  <div
    class="react-switch-bg"
    style="height:13px;width:37px;margin:0;position:relative;background:#aabbcc;border-radius:6.5px;cursor:default;-webkit-transition:background 0.25s;-moz-transition:background 0.25s;transition:background 0.25s"
  >
    <div
      style="height:13px;width:19.5px;position:relative;opacity:1;pointer-events:none;-webkit-transition:opacity 0.25s;-moz-transition:opacity 0.25s;transition:opacity 0.25s"
    >
      <div
        style="color:pink;padding-top:2px;padding-left:6px;font-size:20px"
      >
        ❤
      </div>
    </div>
    <div
      style="height:13px;width:19.5px;position:absolute;opacity:0;right:0;top:0;pointer-events:none;-webkit-transition:opacity 0.25s;-moz-transition:opacity 0.25s;transition:opacity 0.25s"
    >
      <svg
        height="100%"
        style="position:absolute;top:0"
        viewBox="-2 -5 14 20"
        width="100%"
      >
        <path
          d="M9.9 2.12L7.78 0 4.95 2.828 2.12 0 0 2.12l2.83 2.83L0 7.776 2.123 9.9 4.95 7.07 7.78 9.9 9.9 7.776 7.072 4.95 9.9 2.12"
          fill="#fff"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
  <div
    class="react-switch-handle"
    style="height:9px;width:9px;background:#ffffff;display:inline-block;cursor:default;border-radius:50%;position:absolute;transform:translateX(26px);top:2px;outline:0;box-shadow:0px 1px 5px rgba(0, 0, 0, 0.6);border:0;-webkit-transition:background-color 0.25s, transform 0.25s, box-shadow 0.15s;-moz-transition:background-color 0.25s, transform 0.25s, box-shadow 0.15s;transition:background-color 0.25s, transform 0.25s, box-shadow 0.15s"
  />
  <input
    aria-checked="true"
    aria-label="bar"
    aria-labelledby="baz"
    checked=""
    disabled=""
    id="foo"
    role="switch"
    style="border:0;clip:rect(0 0 0 0);height:1px;margin:-1px;overflow:hidden;padding:0;position:absolute;width:1px"
    type="checkbox"
  />
</div>
`;

exports[`Props matches snapshot with default props 1`] = `
<div
  style="position:relative;display:inline-block;text-align:left;opacity:1;direction:ltr;border-radius:14px;-webkit-transition:opacity 0.25s;-moz-transition:opacity 0.25s;transition:opacity 0.25s;touch-action:none;-webkit-tap-highlight-color:rgba(0, 0, 0, 0);-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none"
>
  <div
    class="react-switch-bg"
    style="height:28px;width:56px;margin:0;position:relative;background:#888888;border-radius:14px;cursor:pointer;-webkit-transition:background 0.25s;-moz-transition:background 0.25s;transition:background 0.25s"
  >
    <div
      style="height:28px;width:30px;position:relative;opacity:0;pointer-events:none;-webkit-transition:opacity 0.25s;-moz-transition:opacity 0.25s;transition:opacity 0.25s"
    >
      <svg
        height="100%"
        style="position:absolute;top:0"
        viewBox="-2 -5 17 21"
        width="100%"
      >
        <path
          d="M11.264 0L5.26 6.004 2.103 2.847 0 4.95l5.26 5.26 8.108-8.107L11.264 0"
          fill="#fff"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <div
      style="height:28px;width:30px;position:absolute;opacity:1;right:0;top:0;pointer-events:none;-webkit-transition:opacity 0.25s;-moz-transition:opacity 0.25s;transition:opacity 0.25s"
    >
      <svg
        height="100%"
        style="position:absolute;top:0"
        viewBox="-2 -5 14 20"
        width="100%"
      >
        <path
          d="M9.9 2.12L7.78 0 4.95 2.828 2.12 0 0 2.12l2.83 2.83L0 7.776 2.123 9.9 4.95 7.07 7.78 9.9 9.9 7.776 7.072 4.95 9.9 2.12"
          fill="#fff"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
  <div
    class="react-switch-handle"
    style="height:26px;width:26px;background:#ffffff;display:inline-block;cursor:pointer;border-radius:50%;position:absolute;transform:translateX(1px);top:1px;outline:0;border:0;-webkit-transition:background-color 0.25s, transform 0.25s, box-shadow 0.15s;-moz-transition:background-color 0.25s, transform 0.25s, box-shadow 0.15s;transition:background-color 0.25s, transform 0.25s, box-shadow 0.15s"
  />
  <input
    aria-checked="false"
    role="switch"
    style="border:0;clip:rect(0 0 0 0);height:1px;margin:-1px;overflow:hidden;padding:0;position:absolute;width:1px"
    type="checkbox"
  />
</div>
`;

exports[`Props matches snapshot with horizontal switch 1`] = `
<div
  class="qux"
  style="position:relative;display:inline-block;text-align:left;opacity:0.5;direction:ltr;border-radius:18.5px;-webkit-transition:opacity 0.25s;-moz-transition:opacity 0.25s;transition:opacity 0.25s;touch-action:none;-webkit-tap-highlight-color:rgba(0, 0, 0, 0);-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none"
>
  <div
    class="react-switch-bg"
    style="height:37px;width:13px;margin:0;position:relative;background:#aabbcc;border-radius:18.5px;cursor:default;-webkit-transition:background 0.25s;-moz-transition:background 0.25s;transition:background 0.25s"
  >
    <div
      style="height:19.5px;width:13px;position:relative;opacity:1;pointer-events:none;-webkit-transition:opacity 0.25s;-moz-transition:opacity 0.25s;transition:opacity 0.25s"
    >
      <div
        style="color:pink;padding-top:2px;padding-left:6px;font-size:20px"
      >
        ❤
      </div>
    </div>
    <div
      style="height:19.5px;width:13px;position:absolute;opacity:0;right:0;bottom:0;pointer-events:none;-webkit-transition:opacity 0.25s;-moz-transition:opacity 0.25s;transition:opacity 0.25s"
    >
      <svg
        height="100%"
        style="position:absolute;top:0"
        viewBox="-2 -5 14 20"
        width="100%"
      >
        <path
          d="M9.9 2.12L7.78 0 4.95 2.828 2.12 0 0 2.12l2.83 2.83L0 7.776 2.123 9.9 4.95 7.07 7.78 9.9 9.9 7.776 7.072 4.95 9.9 2.12"
          fill="#fff"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
  <div
    class="react-switch-handle"
    style="height:9px;width:9px;background:#ffffff;display:inline-block;cursor:default;border-radius:50%;position:absolute;transform:translateY(26px);top:0;left:2px;outline:0;box-shadow:0px 1px 5px rgba(0, 0, 0, 0.6);border:0;-webkit-transition:background-color 0.25s, transform 0.25s, box-shadow 0.15s;-moz-transition:background-color 0.25s, transform 0.25s, box-shadow 0.15s;transition:background-color 0.25s, transform 0.25s, box-shadow 0.15s"
  />
  <input
    aria-checked="true"
    aria-label="bar"
    aria-labelledby="baz"
    checked=""
    disabled=""
    id="foo"
    role="switch"
    style="border:0;clip:rect(0 0 0 0);height:1px;margin:-1px;overflow:hidden;padding:0;position:absolute;width:1px"
    type="checkbox"
  />
</div>
`;
